File: ScalaPatterns/src/main/scala/behavioral/ChainOfResponsibility.scala
package behavioral

/**
 * The chain of responsibility pattern decouples the sender of a request from its receiver,
 * by giving more than one object an opportunity to handle the request.
 * The request is processed by the chain until some object handles it.
 *
 * Scala provides a straightforward approach, based on partial functions
 * that are executed only if the argument matches a certain condition
 *
 * @author Daniel Leon
 */
object ChainOfResponsibility {

  case class Event(source: String)

  type EventHandler = PartialFunction[Event, Unit]

  val defaultHandler: EventHandler = PartialFunction(e => println("unknown event " + e.source))

  val keyboardHandler: EventHandler = {
    case Event("keyboard") => println("Display on monitor a character")
  }

  def mouseHandler(delay: Int): EventHandler = {
    case Event("mouse") => println("Move the mouse with delay " + delay)
  }

  def main(args : Array[String]) {

    val keyEvent = Event("keyboard")
    val mouseEvent = Event("mouse")

    val eventHandler = keyboardHandler
      .orElse(mouseHandler(100))
      .orElse(defaultHandler)

    eventHandler(keyEvent)
    eventHandler(mouseEvent)
    eventHandler(Event("bla"))
  }
}

File: JavaPatterns/src/main/java/behavioral/java7/ChainOfResponsibility.java
package behavioral.java7;

/**
 * The chain of responsibility pattern decouples the sender of a request from its receiver,
 * by giving more than one object an opportunity to handle the request.
 * The request is processed by the chain until some object handles it.
 *
 * @author Daniel Leon
 */
interface Event {
    String getSource();
}

class KeyboardEvent implements Event {

    @Override
    public String getSource() {
        return "keyboard";
    }
}

class MouseEvent implements Event {

    @Override
    public String getSource() {
        return "mouse";
    }
}

abstract class EventHandler {
    private EventHandler next;

    void setNext(EventHandler handler) { next = handler; }

    public void handle(Event event) {
        if (canHandle(event)) doHandle(event);
        else if (next != null) next.handle(event);
    }

    abstract protected boolean canHandle(Event event);
    abstract protected void doHandle(Event event);
}

class KeyboardHandler extends EventHandler {

    @Override
    protected boolean canHandle(Event event) {
        return "keyboard".equals(event.getSource());
    }

    @Override
    protected void doHandle(Event event) {
        System.out.println("Display on monitor a character");
    }
}

class MouseHandler extends EventHandler {

    @Override
    protected boolean canHandle(Event event) {
        return "mouse".equals(event.getSource());
    }

    @Override
    protected void doHandle(Event event) {
        System.out.println("Move the mouse");
    }
}

class UndefinedHandler extends EventHandler {

    @Override
    protected boolean canHandle(Event event) {
        return true;
    }

    @Override
    protected void doHandle(Event event) {
        System.out.println("Unknown  event " + event.getSource());
    }
}

public class ChainOfResponsibility {

    public static void main(String[] args) {

        Event keyboardEvent = new KeyboardEvent();
        Event mouseEvent = new MouseEvent();
        Event undefinedEvent = new Event() {
            @Override
            public String getSource() {
                return "undefined";
            }
        };

        KeyboardHandler handler = new KeyboardHandler();
        MouseHandler mouseHandler = new MouseHandler();
        UndefinedHandler undefinedHandler = new UndefinedHandler();

        handler.setNext(mouseHandler);
        mouseHandler.setNext(undefinedHandler);


        handler.handle(keyboardEvent);
        handler.handle(mouseEvent);
        handler.handle(undefinedEvent);
    }
}
